Add setElementCollidableWith server-side

Proxy-99 opened the pull request:
#3681

test code
```lua
vehicle1 = createVehicle ( 411, 2044.65942, 1526.35352, 10.67188)
object = createObject( 3578, 2044.40405, 1510.24036, 10.67188)

setTimer(function()
  res = setElementCollidableWith(vehicle1,object,false)
  print(res)
end, 2000, 1)

```

Proxy-99 committed:
add function serverside

Proxy-99 committed:
element checks

FileEX commented on Client/mods/deathmatch/logic/rpc/CElementRPCs.cpp:
```diff
@@ -765,3 +766,15 @@ void CElementRPCs::SetElementOnFire(CClientEntity* pSource, NetBitStreamInterfac
+void CElementRPCs::SetElementCollidableWith(CClientEntity* pSource, NetBitStreamInterface& bitStream)
+{
+    ElementID ElementID;
+    bool      bCollidable;
```
Hungarian notation

FileEX commented on Client/mods/deathmatch/logic/rpc/CElementRPCs.cpp:
```diff
@@ -765,3 +766,15 @@ void CElementRPCs::SetElementOnFire(CClientEntity* pSource, NetBitStreamInterfac
+    bool      bCollidable;
+
+    bitStream.Read(ElementID);
+    bitStream.ReadBit(bCollidable);
```
bitStream.Read can return false

TheNormalnij replied:
It cannot. The protocol guarantees that the message is fully received.

FileEX replied:
> It cannot. The protocol guarantees that the message is fully received.

So what is the purpose of this function being a bool, and why does most of the code check if it returns true?

TheNormalnij replied:
Code == documentation
It has sense on the server side. The client reads server packets only.
Checks on client could be used to add a new data to a packet without increasing bitstream version, i think

FileEX commented on Client/mods/deathmatch/logic/rpc/CElementRPCs.cpp:
```diff
@@ -765,3 +766,15 @@ void CElementRPCs::SetElementOnFire(CClientEntity* pSource, NetBitStreamInterfac
+    bitStream.Read(ElementID);
+    bitStream.ReadBit(bCollidable);
+
+   CClientEntity* pCollidableWith = CElementIDs::GetElement(ElementID);
```
Hungarian notation + GetElement can return nullptr

FileEX commented on Client/mods/deathmatch/logic/rpc/CElementRPCs.cpp:
```diff
@@ -765,3 +766,15 @@ void CElementRPCs::SetElementOnFire(CClientEntity* pSource, NetBitStreamInterfac
+    bitStream.ReadBit(bCollidable);
+
+   CClientEntity* pCollidableWith = CElementIDs::GetElement(ElementID);
+   pSource->SetCollidableWith(pCollidableWith, bCollidable);
```
```cpp
pSource->SetCollidableWith(pCollidableWith, bitStream.ReadBit());
```

FileEX commented on Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:
```diff
@@ -12566,3 +12565,39 @@ bool CStaticFunctionDefinitions::SpawnVehicleFlyingComponent(CVehicle* const veh
 }
+
+bool CStaticFunctionDefinitions::SetElementCollidableWith(CElement* element, CElement* withElement, bool canCollide)
+    {
```
Indention

Proxy-99 committed:
review fixes

TheNormalnij commented:
The function doesn't work for a new  connected players and after reconnect

FileEX commented on Client/mods/deathmatch/logic/rpc/CElementRPCs.cpp:
```diff
@@ -765,3 +766,19 @@ void CElementRPCs::SetElementOnFire(CClientEntity* pSource, NetBitStreamInterfac
+
+   CClientEntity* collidableWith = CElementIDs::GetElement(ElementID);
+
+   if (collidableWith == nullptr)            // validity check  
+       return;
```
```cpp
if (!collidableWith)
    return;
```

Proxy-99 commented:
> The function doesn't work for a new connected players and after reconnect
 
What packet file should I write the info?
is it   `Server\mods\deathmatch\logic\packets\CPlayerJoinCompletePacket.cpp`

FileEX commented:
> > The function doesn't work for a new connected players and after reconnect
> 
> What packet file should I write the info? is it `Server\mods\deathmatch\logic\packets\CPlayerJoinCompletePacket.cpp`

I think you should store informations on the server side and then send it in the CMapInfoPacket. Check how it's done, for example, with setJetpackWeaponEnabled (CGame::SetJetpackWeaponEnabled & CGame::GetJetpackWeaponEnabled)

Proxy-99 committed:
Merge branch 'master' into setCollide

Proxy-99 committed:
Merge branch 'master' into setCollide

Proxy-99 committed:
Merge branch 'master' into setCollide

Proxy-99 committed:
Merge branch 'master' into setCollide

Proxy-99 committed:
adding store data

need to fix elements not loaded to client before calling setColide function or cause crash nullptr since elements does not created to client yet

Proxy-99 committed:
put it at correct packet

Proxy-99 committed:
Merge branch 'master' into setCollide

Proxy-99 committed:
Merge branch 'master' into setCollide

Proxy-99 committed:
Merge branch 'master' into setCollide

Proxy-99 committed:
Merge branch 'master' into setCollide

Proxy-99 committed:
Merge branch 'master' into setCollide

Proxy-99 committed:
fix strean file

Proxy-99 committed:
follow up version bitstream fix

Proxy-99 committed:
Merge branch 'master' into setCollide
